### DISCUSSION TITLE
Window.Destroy: Await async cleanup

### DISCUSSION THREAD

**Discussion by LukeTOBrien** [dotnet/maui#16510] [Q&A]
Hello,

I would await to await async cleanup operation, these async opperations are requests made to API controllers on the server.  
So in my app I lock records on the database when performing updates so that no one else can modify and I would like to release on app close.  
Also I want to log the user out by making another HTTP request.

In the normal flow, the user would click the 'Logout' button, but the user might click the 'X' close button on Windows or swipe close on Android.

This is the code I have, but I think the Window is destoying itself to quickly for the async opperations to finish:

```cs
protected override Window CreateWindow(IActivationState activationState)
 {
        var window = base.CreateWindow(activationState);
        window.Destroying += Window_Destroying;
        ...

private async void Window_Destroying(object sender, EventArgs e)
    {
        await lockService.ReleaseAllRecordLocks();
        await userService.Unload(false); // <- Log out the user
    }
```

Do you understand what I am trying to do?  
Is this the correct place to have the code?

Thanks,

Luke

**Comment by Xyncgas** [comment]
change ```private async void Window_Destroying(object sender, EventArgs e)``` to ```private async Task```
.NET now supports async through ```Task``` api, async void is not (maybe never in previous .NET versions) monitored they just kinda execute without letting people wait for them

regardless
>This is the code I have, but I think the Window is destoying itself to quickly for the async opperations to finish:

worry not, anything in async void WILL complete, but windows will close before it's completed, if you don't want this behavior, you can try replacing async void with async Task, if that doesn't work which is possible because everything is, you might consider executing this via other method like instead of via window.Destroying maybe execute your function first then call .Dispose on window

Rest assure there are many ways to make sure you execute the function before the window is destroyed you just have to test the behavior of the system you are using and be creative, for example if ```window.Destroying ``` will make sure your function is completed its execution (non async) before window is destoryed, you can put some kind of locks that will not release so ```window.Destroying ``` will have to wait for it
